kaleidoscopic-kings/backend/main.py

```diff
@@ -0,0 +1,176 @@
+        self._timeout_event_cards = {}
+
+    def _build_deck(self, all_cards: List[Card]):
+        for card in all_cards:
+            if card.card_type == self.EVENT:
+                self._event_cards.add(card)
+            elif card.card_type == self.RESPONSE:
+                self._response_cards[card.card_id] = card
+            elif card.card_type == self.START:
+                self._start_card.append(card)
+            elif card.card_type == self.END_GAME:
+                self._game_over_card.append(card)
+
+        self._check_valid_deck()
+
+    def _check_valid_deck(self):
+        if not self._event_cards:
+            raise Exception("No event cards found!")
+        elif len(self._start_card) != 1:
+            raise Exception("There needs to be exactly 1 start card.")
+        elif len(self._game_over_card) != 1:
+            raise Exception("There needs to be exactly 1 game over card.")
+
+    def get_card(self, game_state: GameState) -> Card:
+        """
+        :param game_state: current GameState object. Needed to check current game states with card
+                           conditions.
+        :return: Card object that is valid based on it's conditions and current game state.
+                 When card is drawn it is put
+                 If there are no cards left to draw
+        """
+
+        # Response cards are set as helper variable. If it isn't random then it's a response card.
+        if self._next_card_id != "random":
+            self._reduce_timeouts()
+            return self._response_cards[self._next_card_id]
+
+        possible_event_cards = self._event_cards - self._timeout_event_cards.keys()
+        valid_event_cards = {card for card in possible_event_cards if
+                             card.is_valid(game_state)}
+
+        if len(valid_event_cards) > 0:
+            card = random.choice(tuple(valid_event_cards))
+        else:
+            # There are no more cards to choose from the deck as they are either invalid or most
+            # of them have already been used and are in time-out.
+            # So we choose next card from timed-out cards with the smallest timeout.
+            valid_timeout_cards = {card for card in self._timeout_event_cards if
+                                   card.is_valid(game_state)}
+            card = min(valid_timeout_cards, key=self._timeout_event_cards.get)
+            logger.info(f"No more cards to draw - drawing from timed out cards. "
+                        f"Drawn card {card}")
+            del self._timeout_event_cards[card]
+
+        self._add_card_to_timeout(card)
+        self._reduce_timeouts()
+        return card
+
+    def get_start_card(self) -> Card:
+        return self._start_card[0]
+
+    def get_game_over_card(self) -> Card:
+        return self._game_over_card[0]
+
+    def set_next_response_card(self, card_id: str):
+        self._next_card_id = card_id
```

Comment:
These type of "empty setters" are typically considered non-idiomatic in Python. The Python data model, with its descriptors in the form of `@property`, make them unnecessary. This ensures that we don't need custom setter names and we can write classes that follow the "Uniform access principle" (Meyer, 1988/1997, in *Object-Oriented Software Construction*). For a more in-depth discussion, see chapter 19 and, especially, chapter 20 of Fluent Python.